xlators/mgmt/glusterd/src/glusterd-op-sm.c

```diff
@@ -2973,12 +3105,23 @@ glusterd_op_set_volume(dict_t *dict, char **errstr)
         if (global_opt) {
             cds_list_for_each_entry(voliter, &priv->volumes, vol_list)
             {
+                /* Check dependency of options before setting them */
+                ret = glusterd_dependency_chain_check(voliter, key, value,
+                                                      errstr);
+                if (ret)
+                    goto out;
+
                 value = gf_strdup(value);
                 ret = dict_set_dynstr(voliter->dict, key, value);
                 if (ret)
                     goto out;
             }
         } else {
+            /* Check dependency of options before setting them */
+            ret = glusterd_dependency_chain_check(volinfo, key, value, errstr);
+            if (ret)
+                goto out;
+
```

Comment:
Yeah, but is there any benefit to doing the other way?
If I pass the dict and then go through the options in dict, I will again have to go through the options in a loop, and the code will be repetitive. Mostly, similar to what is being done in the `glusterd_op_set_volume` function.

Reply:
Yeah, the way you did it (and it's still the same in the most recent version as of writing this) is logically faulty.

Imagine we have two options (_optA_ and _optB_) so that either both of them should be turned on, or both of them should be turned off (I'm not sure how much sense does this make  practically, but it's a logical possibility). That is, a scenario that can be expressed with the

```
{optA: 1, optB: 0},
{optA: 0, optB: 1}
```

prohibited clauses.

Now, if both options are off, and dep check is enforced, turning them on sequentially, one first, and then the other, would not be possible,  because you'd need to pass through a prohibited intermediate state (albeit the intended final state would be acceptable). Luckily, `glusterd_op_set_volume` does support them being set simultaneously, via receiving a dict that contains both _optA_ and _optB_, and the cli is also able to leverage this feature, by passing multiple pairs to `gluster vol set` (or using profiles).

However, as of your code, the dep check routine checks on the incoming options one by one. As far as it is concnerned
- `gluster vol set VOL optA on && gluster vol set VOL optB on` and
-  `gluster vol set VOL optA on optB on`

are regarded the same and it will fail the latter form as well.

Reply:
Speaking of this, test case should be added to check the above scenario. What  can I think of some test options could be introduced (using a specific namespace, eg.  `testing.foo`, and then add a few clauses involving them to `prohibited_clauses`  and then various invocations of `gluster vol set` could be put under testing.

However, as `prohibited_clauses` is a single static global entity (and I think that's  fine), if we do this without further ado, the testing clauses will be always matched, even on a prod deployment. So maybe include the testing clauses conditionally for debug builds.

Reply:
As for the repetition: probably that can be handled with a pinch of abstraction. Maybe you could introduce an interator routine for dicts adhering to the `{"key1": <name1>, "value1": <val1>, "key2": <name2>, "value2": <val2>,  ...}` scheme?